New API for Handling Interactive Options in Prompt

T-TonyLiang opened the pull request:
Issue: https://github.com/Shopify/cli-ui/issues/14

### What is accomplished in this PR
- changing the API for handling interactive options, so that they are defined within a block of the `ask` method
- removed the `options` key in the `ask` API
- added a new class to represent the various options and their attached handling procedures
- changed documentation and modified tests

New usage:
```ruby
CLI::UI::Prompt.ask("What is your choice?") do |handler|
    handler.add_option("Choice 1") { |c| puts "Choice 1" }
    handler.add_option("Choice 2") { |c| puts "Choice 2" }
end
```
 
cc @jules2689 @burke for review

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,26 +49,29 @@ class << self
-        # Question with answers
-        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
+        # Multiple choice question with defined handlers
+        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python)) do |handler|
```
I think `add_option` would make `options` redundant

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,26 +49,29 @@ class << self
         #
-        def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true)
-          if (default && !allow_empty) || (options && (default || is_file))
+        def ask(question, default: nil, is_file: nil, allow_empty: true, &define_options)
```
I think we would want to keep `options` and allow options _or_ a block to be given. The block can be used for handlers and the options can be used for returning the option. This also makes this change non-breaking.

jules2689 replied:
e.g.
```
a.ask('My Question') do |handler|
  handler.add_option('A') { method }
  handler.add_option('B') { method_2 }
end
# Choosing 'A' would return the evaluation of `method`
```
but 
```
a.ask('My Question', options: %w(A B))
# Choosing 'A' would return the option A
```

This keeps a nice UI and UX for the code regardless of the choice you want to make. Sometimes you want the former, sometimes the latter.

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,26 +49,29 @@ class << self
+            handler = OptionsHandler.new
+
+            yield handler
+            raise(ArgumentError, 'insufficient options') if handler.options.size < 2
```
Why is 1 option insufficient?

burke replied:
Multiple choice from 1 option? Even then you probably want a yes/no confirm.

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -25,10 +25,11 @@ class << self
         # * +:default+ - The default answer to the question (e.g. they just press enter and don't input anything)
         # * +:is_file+ - Tells the input to use file auto-completion (tab completion)
         # * +:allow_empty+ - Allows the answer to be empty
+        # * +:define_options+ - A Proc that takes a +OptionsHandler+ and uses the +:add_option+ method to
```
This makes it seem like a keyword when we should provide a new section like so:

```
=== Block (optional)
A Proc that takes an OptionsHandler...
```

burke commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,26 +49,29 @@ class << self
-        # Question with answers
-        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
+        # Multiple choice question with defined handlers
+        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python)) do |handler|
```
It looks like `options:` is supposed to go away, right?

T-TonyLiang replied:
Yep! You're right haha, I missed that.

burke commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,26 +49,29 @@ class << self
         #
-        def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true)
-          if (default && !allow_empty) || (options && (default || is_file))
+        def ask(question, default: nil, is_file: nil, allow_empty: true, &define_options)
```
I think leave out the `&define_options`, it's inferred and people generally expect it to be referred to by name if it's listed in the signature.

burke commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,26 +49,29 @@ class << self
-          if options
-            resp = InteractiveOptions.call(options)
+          if block_given?
+            require 'cli/ui/prompt/options_handler'
```
We normally autoload everything within these gems

jules2689 replied:
![image](https://user-images.githubusercontent.com/3074765/34583217-eb057582-f164-11e7-8b70-8db7acaeb1f8.png)

T-TonyLiang replied:
It seemed like `autoload` is being deprecated? (although this was posted around 6 years ago) https://bugs.ruby-lang.org/issues/5653

burke replied:
The reason it's being deprecated has to do with race conditions in multithreaded modes (not typically a major concern in CLI UIs). Even it it does go away, it's still a useful feature and we can reimplement it easily enough.

burke replied:
More importantly, the project is currently autoloading everything. We can have a discussion about whether to switch to late-requires, but let's do that in a different PR: just follow the convention for now.

T-TonyLiang commented:
Updates:
- Addressed the comments
- I wrote a new method called `ask_interactive` which asks the multiple choice questions. There wasn't any reusable code between `ask` for a multiple choice versus a free form question and`ask` was supporting a lot keyword (and block) parameters.
- Additional test cases :)

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,37 +48,20 @@ class << self
-        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
-        #
-        #
         def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true)
```
Talking with @lugray we can both sympathize with the thought that `ask` is a bit of a big method. This being said, we want `ask` to remain the public API. Sometimes when you're writing a public API the capabilities of a method may become a bit big, so we sacrifice a bit of single responsibility to accommodate what people think a method will do instead.

In this case, I don't think we should deprecate `options` and instead we should write something like this:
```ruby
def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true, &options_block)
  if default && !allow_empty
    raise(ArgumentError, 'Cannot combine default and allow_empty keywords')
  end

  has_options = options || options_block
  if has_options && (default || is_file)
    raise(ArgumentError, 'Cannot combine default or is_file with explicit options')
  end

  return ask_with_options(..., &options_block) if has_options
  ask_free_form(...)
end

private

def ask_with_options(...)
end

def ask_free_form(...)
end
```

With this in place, we can call "ask" anywhere, this means it does what you always expect it will be able to do. This does sacrifice a bit of nice code structure for a nicer public API, which is why we split it up entirely on the inside.

Does this make sense?

T-TonyLiang replied:
That definitely makes sense! Thanks

jules2689 commented on lib/cli/ui/prompt/options_handler.rb:
```diff
@@ -0,0 +1,24 @@
+          @options.keys
+        end
+
+        def add_option(option, &handler)
```
This should probably also be just `option`. Adding a verb is generally not preferred for stylistic reasons in Ruby, just a conventions thing. I think we should add a `alias_method :option, :add_option` if you think `add_option` should stay.

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -5,12 +5,13 @@ module CLI
       class << self
-        # Ask a user a question with either free form answer or a set of answers
+        # Ask a user a question with either free form answer or a set of answers (multiple choice)
+        # Can use arrows, y/n, numbers (1/2), and vim bindings to control
```
`to control multiple choice` ?

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -25,14 +26,20 @@ class << self
         #
         # Note:
-        # * +:options+ conflicts with +:default+ and +:is_file+, you cannot set options with either of these keywords
+        # * +:options+ or providing a +Proc+ conflicts with +:default+ and +:is_file+, you cannot set options with either of these keywords
```
It's know as a `block` in this context

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,35 +55,47 @@ class << self
-        # Question with answers
-        #   CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
+        # Interactive (multiple choice) question
+        #   CLI::UI::Prompt.ask_interactive('What kind of project is this?', options: %w(rails go ruby python))
```
`ask` not `ask_interactive`

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -25,14 +26,20 @@ class << self
+        #
+        # ==== Block (optional)
+        #
+        # * A Proc that takes a +OptionsHandler+ and uses the public +:option+ method to add options and their
```
`A block that provides an +OptionsHandler+` sounds a bit better

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,35 +55,47 @@ class << self
+        #   CLI::UI::Prompt.ask_interactive('What kind of project is this?', options: %w(rails go ruby python))
         #
+        # Interactive (multiple choice) question with defined handlers
+        #   CLI::UI::Prompt.ask_interactive('What kind of project is this?', options: %w(rails go ruby python)) do |handler|
```
`ask` here too

jules2689 replied:
We should split it up to:
```ruby
CLI::UI::Prompt.ask('What kind of project is this?', options: %w(rails go ruby python))
```

and 

```ruby
CLI::UI::Prompt.ask('What kind of project is this?') do |handler|
  handler.options('A') { }
end
```

Also `puts selection` in the blocks below returns nil, not selection ;)

T-TonyLiang replied:
Ah, I meant just "outputs selection", but i see the `=>` notation is confusing.

jules2689 commented on lib/cli/ui/prompt.rb:
```diff
@@ -48,35 +55,47 @@ class << self
-            raise(ArgumentError, 'conflicting arguments')
-          end
+        def ask(question, options: nil, default: nil, is_file: nil, allow_empty: true, &options_proc)
+          raise(ArgumentError, 'conflicting arguments') if ((options || block_given?) && (default || is_file))
```
it would be nice to have a better error message now like `explicit options conflicts with default and is_file`
